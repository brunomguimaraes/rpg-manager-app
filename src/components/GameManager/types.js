// @flow

import { NEW_GAME } from "./constants";
import type { RouterInjectedProps } from "~/types/router";
import type { ThunkAction } from "~/types/redux";

export type GameData = {
  name: string,
  sumary: string,
  createdAt: ?Date
};

export type Game = GameData & { id: ?number };

export type NewGame = () => {
  type: typeof NEW_GAME
};

export type FieldValue = { field: string, value: string | ?Date };

export type ChangeGame = (
  fieldValue: FieldValue
) => {
  type: typeof NEW_GAME,
  payload: FieldValue
};

export type GetGame = (id: number) => ThunkAction;
export type SaveGame = (game: GameData) => ThunkAction;
export type UpdateGame = (game: GameData, id: number) => ThunkAction;
export type DeleteGame = (id: number) => ThunkAction;

export type GameManagerProps = RouterInjectedProps & {
  game: Game,
  getGame: GetGame,
  newGame: NewGame,
  saveGame: SaveGame,
  updateGame: UpdateGame,
  deleteGame: DeleteGame,
  changeGame: ChangeGame
};

export type State = {
  game: Game,
  error: boolean,
  loading: boolean
};

export type SaveGameActionTypes =
  | { type: "SAVE_GAME_PENDING" }
  | { type: "SAVE_GAME_SUCCESS" }
  | { type: "SAVE_GAME_FAILURE" };

export type UpdateGameActionTypes =
  | { type: "UPDATE_GAME_PENDING" }
  | { type: "UPDATE_GAME_SUCCESS" }
  | { type: "UPDATE_GAME_FAILURE" };

export type DeleteGameActionTypes =
  | { type: "DELETE_GAME_PENDING" }
  | { type: "DELETE_GAME_SUCCESS" }
  | { type: "DELETE_GAME_FAILURE" };

export type FetchGameActionTypes =
  | { type: "FETCH_GAME_PENDING" }
  | { type: "FETCH_GAME_SUCCESS" }
  | { type: "FETCH_GAME_FAILURE" };

export type Payload = FieldValue | empty; // TODO Add Game to this typecheck

export type ChangeFormAction = { type: "CHANGE_GAME_FORM", payload: Payload };
export type NewGrmeAction = { type: "NEW_GAME", payload: Payload };
export type SaveGameAction = { type: SaveGameActionTypes, payload: Payload };
export type UpdateGameAction = {
  type: UpdateGameActionTypes,
  payload: Payload
};
export type FetchGameAction = { type: FetchGameActionTypes, payload: Payload };
export type DeleteGameAction = {
  type: DeleteGameActionTypes,
  payload: Payload
};

export type Action =
  | ChangeFormAction
  | NewGrmeAction
  | SaveGameAction
  | UpdateGameAction
  | FetchGameAction
  | DeleteGameAction;
