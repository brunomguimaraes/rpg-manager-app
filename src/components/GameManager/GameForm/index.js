// @flow

export { default } from "./GameForm";
