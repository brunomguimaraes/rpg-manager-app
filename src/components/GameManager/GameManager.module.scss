@import "~styles/variables";

.gameManager {
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;

  .placeholder {
    width: 80%;
    height: 80%;
    display: flex;
    background: $color4;
    flex-direction: column;
  }

  .header {
    display: flex;
    color: $color4;
    padding: $spacer;
    align-items: center;
    background: $color1;
    height: $spacer * 6;
    box-sizing: border-box;
    justify-content: space-between;
    & > h4 {
      margin: 0;
    }
  }

  .form {
    height: 100%;
    display: flex;
    width: 100%;
    flex-direction: column;
  }
}
