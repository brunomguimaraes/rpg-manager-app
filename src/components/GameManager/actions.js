// @flow

import { createAction } from "redux-actions";
import { history } from "~/utils/browser";
import client from "~/api";
import {
  NEW_GAME,
  FETCH_GAME,
  SAVE_GAME,
  UPDATE_GAME,
  DELETE_GAME,
  CHANGE_GAME_FORM
} from "./constants";
import type { FieldValue } from "./types";

export function changeGame({ field, value }: FieldValue) {
  return {
    type: CHANGE_GAME_FORM,
    payload: { field, value }
  };
}

export function newGame() {
  return { type: NEW_GAME };
}

export const getGame = createAction(
  FETCH_GAME,
  async id => await client.getGame(id)
);

export const saveGame = createAction(SAVE_GAME, async game => {
  try {
    const { id } = await client.saveGame(game);
    history.push(`/game/${id}`);
    history.go();
  } catch (error) {
    console.error(error);
  }
});

export const updateGame = createAction(
  UPDATE_GAME,
  async (game, id) => await client.updateGame(id, game)
);

export const deleteGame = createAction(DELETE_GAME, async id => {
  await client.deleteGame(id);
  history.push("/dashboard");
  history.go();
});
