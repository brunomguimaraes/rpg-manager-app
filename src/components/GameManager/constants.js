// @flow

const NEW_GAME = "NEW_GAME";
const FETCH_GAME = "FETCH_GAME";
const SAVE_GAME = "SAVE_GAME";
const UPDATE_GAME = "UPDATE_GAME";
const DELETE_GAME = "DELETE_GAME";
const CHANGE_GAME_FORM = "CHANGE_GAME_FORM";

export {
  FETCH_GAME,
  SAVE_GAME,
  UPDATE_GAME,
  DELETE_GAME,
  CHANGE_GAME_FORM,
  NEW_GAME
};
