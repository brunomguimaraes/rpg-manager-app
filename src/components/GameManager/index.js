// @flow

import Loadable from "react-loadable";
import Loading from "~/components/Loading";

export default Loadable({
  loader: () => import("./GameManager"),
  loading: Loading
});

export { default as gameManagerReducer } from "./reducers";
export type { Game, Action, State } from "./types";
