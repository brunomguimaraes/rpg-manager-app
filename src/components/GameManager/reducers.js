// @flow

import {
  NEW_GAME,
  CHANGE_GAME_FORM,
  FETCH_GAME,
  SAVE_GAME,
  UPDATE_GAME,
  DELETE_GAME
} from "./constants";
import type { State, Action } from "./types";

export const initialState: State = {
  game: {
    id: null,
    name: "",
    sumary: "",
    createdAt: null
  },
  error: false,
  loading: false
};

function reducer(state: State = initialState, action: Action) {
  const { type, payload } = action;
  switch (type) {
    case NEW_GAME:
      return {
        ...initialState
      };
    case CHANGE_GAME_FORM:
      return {
        ...state,
        game: {
          ...state.game,
          [payload.field]: payload.value
        }
      };
    case `${FETCH_GAME}_PENDING`:
      return {
        ...state,
        error: false,
        loading: true
      };
    case `${FETCH_GAME}_SUCCESS`:
      return {
        error: false,
        loading: false,
        game: payload
      };
    case `${FETCH_GAME}_FAILURE`:
      return {
        ...state,
        error: true,
        loading: false
      };
    case `${SAVE_GAME}_PENDING`:
      return {
        ...state,
        error: false,
        loading: true
      };
    case `${SAVE_GAME}_SUCCESS`:
      return {
        error: false,
        loading: false,
        game: payload
      };
    case `${SAVE_GAME}_FAILURE`:
      return {
        ...state,
        error: true,
        loading: false
      };
    case `${UPDATE_GAME}_PENDING`:
      return {
        ...state,
        error: false,
        loading: true
      };
    case `${UPDATE_GAME}_SUCCESS`:
      return {
        error: false,
        loading: false,
        game: payload
      };
    case `${UPDATE_GAME}_FAILURE`:
      return {
        ...state,
        error: true,
        loading: false
      };
    case `${DELETE_GAME}_PENDING`:
      return {
        ...state,
        error: false,
        loading: true
      };
    case `${DELETE_GAME}_SUCCESS`:
      return { ...initialState };
    case `${DELETE_GAME}_FAILURE`:
      return {
        ...state,
        error: true,
        loading: false
      };
    default:
      return state;
  }
}

export default reducer;
