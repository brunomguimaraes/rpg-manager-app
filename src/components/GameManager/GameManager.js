// @flow

import React, { Component } from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import ContentDelete from "material-ui/svg-icons/action/delete";
import IconButton from "material-ui/IconButton";
import GameForm from "./GameForm";
import {
  changeGame,
  getGame,
  saveGame,
  updateGame,
  newGame,
  deleteGame
} from "./actions";
import { TRANSPARENT1 } from "~/styles/constants";
import styles from "./GameManager.module.scss";
import type { GameManagerProps as Props } from "./types";

export class GameManager extends Component<Props, { tab: string }> {
  unlisten: Function;

  componentWillUnmount() {
    if (this.unlisten) {
      this.unlisten();
    }
  }

  componentDidMount() {
    const { match, getGame } = this.props;
    if (match.params.id) {
      getGame(match.params.id);
    }

    this.unlisten = this.props.history.listen(_ => {
      this.props.newGame();
    });
  }

  render() {
    const { match, ...rest } = this.props;
    const { params } = match;
    const { id } = params;

    return (
      <div className={styles.gameManager}>
        <section className={styles.placeholder}>
          <div className={styles.form}>
            <header className={styles.header}>
              <h4>{id ? "EDIT CAMPAIGN" : "NEW CAMPAIGN"}</h4>
              {id && (
                <IconButton
                  hoveredStyle={{ background: TRANSPARENT1 }}
                  tooltip="Delete this campaign"
                  onClick={() => {
                    rest.deleteGame(id);
                  }}
                >
                  <ContentDelete color="white" />
                </IconButton>
              )}
            </header>
            <GameForm
              {...rest.game}
              id={id}
              onChange={rest.changeGame}
              onSubmit={id ? rest.updateGame : rest.saveGame}
            />
          </div>
        </section>
      </div>
    );
  }
}

function selector({ gameManager }) {
  return { ...gameManager };
}

function dispatcher(dispatch) {
  return bindActionCreators(
    {
      newGame,
      getGame,
      saveGame,
      changeGame,
      deleteGame,
      updateGame
    },
    dispatch
  );
}

export default connect(selector, dispatcher)(GameManager);
