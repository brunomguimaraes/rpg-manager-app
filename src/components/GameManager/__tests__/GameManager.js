import React from "react";
import { GameManager } from "../GameManager";
import { initialState } from "../reducers";

const unlisten = jest.fn();
const props = {
  match: { params: { id: null } },
  history: { listen: jest.fn(() => unlisten) },
  game: initialState.game,
  newGame: jest.fn(),
  getGame: jest.fn(),
  saveGame: jest.fn(),
  changeGame: jest.fn(),
  deleteGame: jest.fn(),
  updateGame: jest.fn()
};
GameManager.prototype.unlisten = jest.fn();

test("GameManager renders with no path id", () => {
  const component = shallow(<GameManager {...props} />);
  expect(component).toMatchSnapshot();
  expect(props.getGame).not.toHaveBeenCalled();
});

test("GameManager renders with path id", () => {
  const newProps = {
    ...props,
    match: { params: { id: 1 } }
  };
  const component = shallow(<GameManager {...newProps} />);
  expect(component).toMatchSnapshot();
  expect(props.getGame).toHaveBeenCalledWith(1);
});

test("GameManager listen to route changes and calls newGame", () => {
  const component = shallow(<GameManager {...props} />);
  const { listen } = props.history;
  expect(listen).toHaveBeenCalled();
  const [callback] = listen.mock.calls[0];
  callback();
  expect(props.newGame).toHaveBeenCalled();
});

test("GameManager unmount", () => {
  const component = shallow(<GameManager {...props} />);
  const instance = component.instance();
  instance.componentWillUnmount();
  expect(instance.unlisten).toHaveBeenCalled();
});
