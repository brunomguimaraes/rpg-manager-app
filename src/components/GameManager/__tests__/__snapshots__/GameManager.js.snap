// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameManager renders with no path id 1`] = `
<div
  className="gameManager"
>
  <section
    className="placeholder"
  >
    <div
      className="form"
    >
      <header
        className="header"
      >
        <h4>
          NEW CAMPAIGN
        </h4>
      </header>
      <GameForm
        createdAt={null}
        id={null}
        name=""
        onChange={[MockFunction]}
        onSubmit={[MockFunction]}
        sumary=""
      />
    </div>
  </section>
</div>
`;

exports[`GameManager renders with path id 1`] = `
<div
  className="gameManager"
>
  <section
    className="placeholder"
  >
    <div
      className="form"
    >
      <header
        className="header"
      >
        <h4>
          EDIT CAMPAIGN
        </h4>
        <IconButton
          disableTouchRipple={false}
          disabled={false}
          hoveredStyle={
            Object {
              "background": "rgba(170, 170, 170, 0.13)",
            }
          }
          iconStyle={Object {}}
          onClick={[Function]}
          tooltip="Delete this campaign"
          tooltipPosition="bottom-center"
          touch={false}
        >
          <ActionDelete
            color="white"
          />
        </IconButton>
      </header>
      <GameForm
        createdAt={null}
        id={1}
        name=""
        onChange={[MockFunction]}
        onSubmit={[MockFunction]}
        sumary=""
      />
    </div>
  </section>
</div>
`;
